reps/2022-03-10-plasma_alternative.md

```diff
@@ -0,0 +1,185 @@
+
+  // get objects by object_ids. add the reference 
+  virtual Status Get(const std::vector<ObjectID> &object_ids, int64_t timeout_ms,
+             std::vector<ObjectBuffer> *object_buffers, bool is_from_worker) = 0;
+
+  // notify the server that this client does not need the object. remove the reference
+  virtual Status Release(const ObjectID &object_id) = 0;
+
+  // check the existence
+  virtual Status Contains(const ObjectID &object_id, bool *has_object) = 0;
+
+  // Abort an unsealed object as if it was never created at all.
+  virtual Status Abort(const ObjectID &object_id) = 0;
+
+  // Seal the object, make it immutable.
+  virtual Status Seal(const ObjectID &object_id) = 0;
+
+  // Delete the objects only when (1) they exist, (2) has been sealed, (3) not used by any clients. or it will do nothing.
+  virtual Status Delete(const std::vector<ObjectID> &object_ids) = 0;
+
+  // Delete objects until we have freed up num_bytes bytes.
+  virtual Status Evict(int64_t num_bytes, int64_t &num_bytes_evicted) = 0;
+
+  // disconnect from the store server.
+  virtual Status Disconnect() = 0;
+
+  // query the capacity of the object store.
+  virtual int64_t store_capacity() = 0;
+};
+```
+
+**Step 2:** **Vineyard implement the interface to provide plamsa-like put/get/delete.** Like Plasma, Vineyard also has client-side and server-side. The client slide provides Plasma-compatible APIs which facilitates us to wrap Plasma requests into Vineyard requests.  We will implement `VineyardClientImpl` in a separate file. Both `PlasmaClientImpl` and `VineyardClientImpl` inherit from the `ClientImplIterface`, which allow users to select the desired client(Plasma or Vineyard) at runtime.
+
+**Step 3:** **Make the behavior of VineyardClientImpl consistent with PlasmaClientImpl.** Vineyard server also provide simlar function like plasma server, it also has reference count, eviction, and lifecycle management. We will try our best to align the behavior of plasma server and vineyard server. We will maintain a special code path in vineyard to provide the same functionality as Plasma. The elusive point maybe the spilling, since it is a callback which requires the raylet resources and cannot be executed in another process. An intuitively solution is to implement a proxy to replace the PlasmaStoreRunner in vineyard mode, which waits the messages from vineyard server to trigger these callback.
+
+**Step 4:** **python/cpp/jave wrappers.** Handle the python/cpp/jave wrappers. provide demos, examples, and docs.
+
+### Example - Code
+
+To share the objects with out-of-Ray systems, vineyard provides API to persist a Ray object into an object that external systems can recognize(e.g. vineyard object in this proposal). The conversion should be zero-copy. The conversion APIs are provided by vineyard.
+```python
+# users get objects from ray then put (zero-copy) them into vineyard or vice versa.
```

Comment:
The put here is to make a ray object visible to other vineyard clients. Since the ray object is already on top of vineyard store in this case, thus the put is zero-copy.